src/v2.rs

```diff
@@ -0,0 +1,954 @@
+            let module = EszipV2Module::Module {
+              kind: ModuleKind::Json,
+              source: EszipV2SourceSlot::Ready(Arc::new(
+                source.as_bytes().to_owned(),
+              )),
+              source_map: EszipV2SourceSlot::Ready(Arc::new(vec![])),
+            };
+            modules.insert(specifier, module);
+          }
+        }
+        _ => {}
+      }
+
+      ordered_modules.push(specifier.to_string());
+      for dep in module.dependencies.values() {
+        if let Some(specifier) = dep.get_code() {
+          visit_module(
+            graph,
+            emit_options,
+            modules,
+            ordered_modules,
+            specifier,
+          )?;
+        }
+      }
+
+      Ok(())
+    }
+
+    for (root, kind) in &graph.roots {
+      assert!(matches!(
+        kind,
+        deno_graph::ModuleKind::Esm | deno_graph::ModuleKind::Asserted
+      ));
+      visit_module(
+        &graph,
+        &emit_options,
+        &mut modules,
+        &mut ordered_modules,
+        root,
+      )?;
+    }
+
+    for (specifier, target) in &graph.redirects {
+      let module = EszipV2Module::Redirect {
+        target: target.to_string(),
+      };
+      modules.insert(specifier.to_string(), module);
+    }
+
+    Ok(Self {
+      modules: Arc::new(Mutex::new(modules)),
+      ordered_modules,
+    })
+  }
+
+  /// Get the module metadata for a given module specifier. This function will
+  /// follow redirects. The returned module has functions that can be used to
+  /// obtain the module source and source map.
+  pub fn get_module(&self, specifier: &str) -> Option<Module> {
+    let mut specifier = specifier;
+    let mut visited = HashSet::new();
+    let modules = self.modules.lock().unwrap();
+    loop {
+      visited.insert(specifier);
+      let module = modules.get(specifier)?;
+      match module {
+        EszipV2Module::Module { kind, .. } => {
+          return Some(Module {
+            specifier: specifier.to_string(),
+            kind: *kind,
+            inner: ModuleInner::V2(EszipV2 {
+              modules: self.modules.clone(),
+              ordered_modules: vec![],
+            }),
+          });
+        }
+        EszipV2Module::Redirect { target } => {
+          specifier = target;
+          if visited.contains(specifier) {
+            return None;
+          }
+        }
+      }
+    }
+  }
+
+  pub(crate) async fn get_module_source<'a>(
+    &'a self,
+    specifier: &str,
+  ) -> Arc<Vec<u8>> {
+    poll_fn(|cx| {
+      let mut modules = self.modules.lock().unwrap();
+      let module = modules.get_mut(specifier).unwrap();
+      let slot = match module {
+        EszipV2Module::Module { source, .. } => source,
+        EszipV2Module::Redirect { .. } => {
+          panic!("redirects are already resolved")
+        }
+      };
+      match slot {
+        EszipV2SourceSlot::Pending { wakers, .. } => {
+          wakers.push(cx.waker().clone());
```

Comment:
This is a valid solution, just wondering if [`shared`](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.shared) might be applicable here?

Reply:
Could also use that, sure. It's really the same thing.